provider/kubernetes/client.go

```diff
@@ -150,129 +122,84 @@ func (c *clientImpl) WatchAll(namespaces Namespaces, labelSelector string, stopC
-		informer := informer
-		safe.Go(func() {
-			wg.Add(1)
-			informer.Run(stopCh)
-			wg.Done()
+		factory := informers.NewFilteredSharedInformerFactory(c.clientset, resyncPeriod, ns, func(opts *metav1.ListOptions) {
+			opts.LabelSelector = labelSelector
 		})
+		factory.Extensions().V1beta1().Ingresses().Informer().AddEventHandler(eventHandler)
+		factory.Core().V1().Services().Informer().AddEventHandler(eventHandler)
+		factory.Core().V1().Endpoints().Informer().AddEventHandler(eventHandler)
+		factory.Core().V1().Secrets().Informer().AddEventHandler(eventHandler)
+		c.factories[ns] = factory
 	}
 
-	if !cache.WaitForCacheSync(stopCh, informManager.syncFuncs...) {
-		return nil, fmt.Errorf("timed out waiting for controller caches to sync")
+	for _, ns := range namespaces {
+		c.factories[ns].Start(stopCh)
 	}
 
-	safe.Go(func() {
-		<-stopCh
-		wg.Wait()
-		close(eventCh)
-	})
-
-	return eventCh, nil
-}
-
-// WatchIngresses sets up a watch on Ingress objects and returns a corresponding shared informer.
-func (c *clientImpl) WatchIngresses(namespace string, labelSelector labels.Selector, watchCh chan<- interface{}) cache.SharedInformer {
-	listOptions := metav1.ListOptions{
-		LabelSelector: labelSelector.String(),
-		FieldSelector: fields.Everything().String(),
+	for _, ns := range namespaces {
+		for t, ok := range c.factories[ns].WaitForCacheSync(stopCh) {
+			if !ok {
+				return nil, fmt.Errorf("timed out waiting for controller caches to sync %s in namespace %q", t.String(), ns)
+			}
+		}
 	}
-	informer := loadInformer(&cache.ListWatch{
-		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-			return c.clientset.ExtensionsV1beta1().Ingresses(namespace).List(listOptions)
-		},
-		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-			return c.clientset.ExtensionsV1beta1().Ingresses(namespace).Watch(listOptions)
-		},
-	}, &extensionsv1beta1.Ingress{}, watchCh)
-	c.ingStores = append(c.ingStores, informer.GetStore())
-	return informer
-}
-
-// WatchObjects sets up a watch on objects and returns a corresponding shared informer.
-func (c *clientImpl) WatchObjects(namespace, kind string, object runtime.Object, storeMap map[string]cache.Store, watchCh chan<- interface{}) cache.SharedInformer {
-	listWatch := cache.NewListWatchFromClient(
-		c.clientset.CoreV1().RESTClient(),
-		kind,
-		namespace,
-		fields.Everything())
-
-	informer := loadInformer(listWatch, object, watchCh)
-	storeMap[namespace] = informer.GetStore()
-	return informer
-}
 
-func loadInformer(listWatch cache.ListerWatcher, object runtime.Object, watchCh chan<- interface{}) cache.SharedInformer {
-	informer := cache.NewSharedInformer(
-		listWatch,
-		object,
-		resyncPeriod,
-	)
+	// Do not wait for the Secrets store to get synced since we cannot rely on
+	// users having granted RBAC permissions for this object.
+	// https://github.com/containous/traefik/issues/1784 should improve the
+	// situation here in the future.
```

Comment:
Is the comment true and correctly placed here? It looks like lines 141 to 142 are including Secrets, or am I missing something?